File: tests/helpers/api-client.js
/**
 * API Client for E2E Testing
 * Provides utilities for making API requests to the license server
 */

// Use node-fetch if available, otherwise use global fetch (Node 18+)
let fetch;
try {
    fetch = require('node-fetch');
} catch (e) {
    // Node 18+ has fetch built-in
    if (typeof globalThis.fetch === 'function') {
        fetch = globalThis.fetch;
    } else {
        throw new Error('fetch is not available. Install node-fetch or use Node.js 18+');
    }
}

const BASE_URL = process.env.LICENSE_SERVER_URL || 'http://localhost:8000';

// Logging utility
const log = {
    enabled: process.env.TEST_LOGGING !== 'false',
    request: (method, url, data) => {
        if (log.enabled) {
            console.log(`[API] ${method} ${url}`, data ? JSON.stringify(data, null, 2) : '');
        }
    },
    response: (status, data) => {
        if (log.enabled) {
            console.log(`[API] Response ${status}`, JSON.stringify(data, null, 2));
        }
    },
    error: (error) => {
        if (log.enabled) {
            console.error(`[API] Error:`, error.message);
        }
    }
};

class ApiClient {
    constructor(baseUrl = BASE_URL, options = {}) {
        this.baseUrl = baseUrl;
        this.sessionToken = null;
        this.timeout = options.timeout || 30000; // 30 seconds default
        this.retries = options.retries || 3;
        this.retryDelay = options.retryDelay || 1000; // 1 second
        this.logging = options.logging !== false;
    }

    /**
     * Make a request with retry logic
     */
    async _requestWithRetry(method, endpoint, body = null, headers = {}) {
        const url = `${this.baseUrl}${endpoint}`;
        let lastError;
        
        for (let attempt = 1; attempt <= this.retries; attempt++) {
            try {
                if (this.logging) {
                    log.request(method, url, body);
                }
                
                const controller = new AbortController();
                const timeoutId = setTimeout(() => controller.abort(), this.timeout);
                
                const requestOptions = {
                    method,
                    headers: {
                        'Content-Type': 'application/json',
                        ...headers,
                    },
                    signal: controller.signal,
                };
                
                if (body) {
                    requestOptions.body = JSON.stringify(body);
                }
                
                const response = await fetch(url, requestOptions);
                clearTimeout(timeoutId);
                
                const responseData = {
                    status: response.status,
                    data: await response.json().catch(() => ({})),
                    headers: Object.fromEntries(response.headers.entries()),
                };
                
                if (this.logging) {
                    log.response(response.status, responseData.data);
                }
                
                // Retry on 5xx errors or network errors
                if (response.status >= 500 && attempt < this.retries) {
                    await this._delay(this.retryDelay * attempt);
                    continue;
                }
                
                return responseData;
            } catch (error) {
                lastError = error;
                if (this.logging) {
                    log.error(error);
                }
                
                // Retry on network errors
                if (attempt < this.retries && (error.name === 'AbortError' || error.code === 'ECONNREFUSED')) {
                    await this._delay(this.retryDelay * attempt);
                    continue;
                }
                
                throw error;
            }
        }
        
        throw lastError || new Error('Request failed after retries');
    }
    
    /**
     * Delay helper for retries
     */
    _delay(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    /**
     * Make a GET request
     */
    async get(endpoint, headers = {}) {
        return this._requestWithRetry('GET', endpoint, null, headers);
    }

    /**
     * Make a POST request
     */
    async post(endpoint, body = {}, headers = {}) {
        return this._requestWithRetry('POST', endpoint, body, headers);
    }
    
    /**
     * Make a DELETE request
     */
    async delete(endpoint, headers = {}) {
        return this._requestWithRetry('DELETE', endpoint, null, headers);
    }
    
    /**
     * Make a PUT request
     */
    async put(endpoint, body = {}, headers = {}) {
        return this._requestWithRetry('PUT', endpoint, body, headers);
    }
    
    /**
     * Make parallel requests
     */
    async parallel(requests) {
        return Promise.all(requests.map(req => {
            if (req.method === 'GET') {
                return this.get(req.endpoint, req.headers);
            } else if (req.method === 'POST') {
                return this.post(req.endpoint, req.body, req.headers);
            } else if (req.method === 'DELETE') {
                return this.delete(req.endpoint, req.headers);
            } else if (req.method === 'PUT') {
                return this.put(req.endpoint, req.body, req.headers);
            }
            throw new Error(`Unsupported method: ${req.method}`);
        }));
    }

    /**
     * Login with email/password and store session token
     */
    async login(email, password, deviceId = 'TEST_DEVICE', deviceName = 'Test Device') {
        const response = await this.post('/api/v2/auth/login', {
            email,
            password,
            device_id: deviceId,
            device_name: deviceName,
        });
        if (response.status === 200 && response.data.session_token) {
            this.sessionToken = response.data.session_token;
        }
        return response;
    }

    /**
     * Login with magic link token
     */
    async loginWithMagicLink(magicLinkToken, deviceId = 'TEST_DEVICE', deviceName = 'Test Device') {
        const response = await this.post('/api/v2/auth/login', {
            magic_link_token: magicLinkToken,
            device_id: deviceId,
            device_name: deviceName,
        });
        if (response.status === 200 && response.data.session_token) {
            this.sessionToken = response.data.session_token;
        }
        return response;
    }

    /**
     * Verify magic link token (alternative endpoint)
     */
    async verifyMagicLink(magicLinkToken, deviceId = 'TEST_DEVICE', deviceName = 'Test Device') {
        return this.loginWithMagicLink(magicLinkToken, deviceId, deviceName);
    }

    /**
     * Refresh session token
     */
    async refresh(deviceId = 'TEST_DEVICE') {
        if (!this.sessionToken) {
            throw new Error('No session token available. Login first.');
        }
        const response = await this.post('/api/v2/auth/refresh', {
            session_token: this.sessionToken,
            device_id: deviceId,
        });
        if (response.status === 200 && response.data.session_token) {
            this.sessionToken = response.data.session_token;
        }
        return response;
    }

    /**
     * Logout (revoke session token)
     */
    async logout() {
        if (!this.sessionToken) {
            throw new Error('No session token available. Login first.');
        }
        const response = await this.post('/api/v2/auth/logout', {
            session_token: this.sessionToken,
        });
        if (response.status === 200) {
            this.sessionToken = null;
        }
        return response;
    }

    /**
     * Health check
     */
    async health() {
        return await this.get('/api/v2/health');
    }

    /**
     * Get license info (requires authentication)
     */
    async getLicenseInfo() {
        if (!this.sessionToken) {
            throw new Error('No session token available. Login first.');
        }
        return await this.get('/api/v2/license/info', {
            'Authorization': `Bearer ${this.sessionToken}`,
        });
    }

    /**
     * Validate license (requires authentication)
     */
    async validateLicense(entitlementToken) {
        if (!this.sessionToken) {
            throw new Error('No session token available. Login first.');
        }
        return await this.post('/api/v2/license/validate', {
            entitlement_token: entitlementToken,
        }, {
            'Authorization': `Bearer ${this.sessionToken}`,
        });
    }

    /**
     * Register device (requires authentication)
     */
    async registerDevice(deviceId, deviceName = 'Test Device') {
        if (!this.sessionToken) {
            throw new Error('No session token available. Login first.');
        }
        return await this.post('/api/v2/devices/register', {
            device_id: deviceId,
            device_name: deviceName,
        }, {
            'Authorization': `Bearer ${this.sessionToken}`,
        });
    }

    /**
     * List devices (requires authentication)
     */
    async listDevices() {
        if (!this.sessionToken) {
            throw new Error('No session token available. Login first.');
        }
        return await this.get('/api/v2/devices', {
            'Authorization': `Bearer ${this.sessionToken}`,
        });
    }

    /**
     * Delete device (requires authentication)
     */
    async deleteDevice(deviceId) {
        if (!this.sessionToken) {
            throw new Error('No session token available. Login first.');
        }
        return this.delete(`/api/v2/devices/${deviceId}`, {
            'Authorization': `Bearer ${this.sessionToken}`,
        });
    }

    /**
     * Clear session
     */
    clearSession() {
        this.sessionToken = null;
    }
}

module.exports = ApiClient;

File: tests/e2e/helpers/api_client.py
"""
API Client Helper for E2E Tests
Provides a wrapper around API requests with retry logic, logging, and error handling
"""

import requests
import time
import logging
from typing import Optional, Dict, Any, Tuple
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry

from tests.e2e.test_config import (
    API_BASE_URL,
    API_RETRY_ATTEMPTS,
    API_RETRY_DELAY,
    TEST_TIMEOUT,
    LOG_LEVEL
)

logger = logging.getLogger(__name__)


class APIClient:
    """API client with retry logic and error handling"""
    
    def __init__(self, base_url: str = API_BASE_URL, timeout: int = TEST_TIMEOUT):
        self.base_url = base_url
        self.timeout = timeout
        self.session = requests.Session()
        self.token: Optional[str] = None
        
        # Configure retry strategy
        retry_strategy = Retry(
            total=API_RETRY_ATTEMPTS,
            backoff_factor=API_RETRY_DELAY,
            status_forcelist=[429, 500, 502, 503, 504],
            allowed_methods=["HEAD", "GET", "POST", "PUT", "DELETE", "OPTIONS", "TRACE"]
        )
        adapter = HTTPAdapter(max_retries=retry_strategy)
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)
        
        # Set default headers
        self.session.headers.update({
            'Content-Type': 'application/json',
            'Accept': 'application/json'
        })
    
    def set_token(self, token: str):
        """Set authentication token"""
        self.token = token
        self.session.headers.update({'Authorization': f'Bearer {token}'})
    
    def clear_token(self):
        """Clear authentication token"""
        self.token = None
        self.session.headers.pop('Authorization', None)
    
    def _log_request(self, method: str, url: str, **kwargs):
        """Log API request"""
        if LOG_LEVEL == 'DEBUG':
            logger.debug(f"{method} {url}")
            if 'json' in kwargs:
                logger.debug(f"Request body: {kwargs['json']}")
    
    def _log_response(self, response: requests.Response):
        """Log API response"""
        if LOG_LEVEL == 'DEBUG':
            logger.debug(f"Response: {response.status_code}")
            try:
                logger.debug(f"Response body: {response.json()}")
            except:
                logger.debug(f"Response body: {response.text[:200]}")
    
    def get(self, endpoint: str, **kwargs) -> requests.Response:
        """GET request"""
        url = f"{self.base_url}/{endpoint.lstrip('/')}"
        self._log_request('GET', url, **kwargs)
        response = self.session.get(url, timeout=self.timeout, **kwargs)
        self._log_response(response)
        return response
    
    def post(self, endpoint: str, **kwargs) -> requests.Response:
        """POST request"""
        url = f"{self.base_url}/{endpoint.lstrip('/')}"
        self._log_request('POST', url, **kwargs)
        response = self.session.post(url, timeout=self.timeout, **kwargs)
        self._log_response(response)
        return response
    
    def put(self, endpoint: str, **kwargs) -> requests.Response:
        """PUT request"""
        url = f"{self.base_url}/{endpoint.lstrip('/')}"
        self._log_request('PUT', url, **kwargs)
        response = self.session.put(url, timeout=self.timeout, **kwargs)
        self._log_response(response)
        return response
    
    def delete(self, endpoint: str, **kwargs) -> requests.Response:
        """DELETE request"""
        url = f"{self.base_url}/{endpoint.lstrip('/')}"
        self._log_request('DELETE', url, **kwargs)
        response = self.session.delete(url, timeout=self.timeout, **kwargs)
        self._log_response(response)
        return response
    
    def health_check(self) -> Tuple[bool, Optional[str]]:
        """Check API health"""
        try:
            response = self.get('/health')
            if response.status_code == 200:
                return True, None
            return False, f"Health check failed: {response.status_code}"
        except Exception as e:
            return False, f"Health check error: {str(e)}"
    
    def login(self, email: str, password: str, device_name: Optional[str] = None) -> Tuple[bool, Optional[Dict[str, Any]], Optional[str]]:
        """
        Login and get session token
        
        Returns:
            (success, response_data, error_message)
        """
        try:
            payload = {
                'email': email,
                'password': password
            }
            if device_name:
                payload['device_name'] = device_name
            
            response = self.post('/auth/login', json=payload)
            
            if response.status_code == 200:
                data = response.json()
                if 'session_token' in data:
                    self.set_token(data['session_token'])
                return True, data, None
            else:
                error_msg = response.json().get('message', f"Login failed: {response.status_code}")
                return False, None, error_msg
        except Exception as e:
            return False, None, f"Login error: {str(e)}"
    
    def refresh_token(self, refresh_token: str) -> Tuple[bool, Optional[Dict[str, Any]], Optional[str]]:
        """Refresh session token"""
        try:
            response = self.post('/auth/refresh', json={'refresh_token': refresh_token})
            
            if response.status_code == 200:
                data = response.json()
                if 'session_token' in data:
                    self.set_token(data['session_token'])
                return True, data, None
            else:
                error_msg = response.json().get('message', f"Token refresh failed: {response.status_code}")
                return False, None, error_msg
        except Exception as e:
            return False, None, f"Token refresh error: {str(e)}"
    
    def logout(self) -> Tuple[bool, Optional[str]]:
        """Logout"""
        try:
            response = self.post('/auth/logout')
            self.clear_token()
            
            if response.status_code == 200:
                return True, None
            else:
                return False, f"Logout failed: {response.status_code}"
        except Exception as e:
            return False, f"Logout error: {str(e)}"
    
    def request_magic_link(self, email: str) -> Tuple[bool, Optional[str]]:
        """Request magic link"""
        try:
            response = self.post('/auth/magic-link/request', json={'email': email})
            
            if response.status_code == 200:
                return True, None
            else:
                error_msg = response.json().get('message', f"Magic link request failed: {response.status_code}")
                return False, error_msg
        except Exception as e:
            return False, f"Magic link request error: {str(e)}"
    
    def verify_magic_link(self, token: str) -> Tuple[bool, Optional[Dict[str, Any]], Optional[str]]:
        """Verify magic link token"""
        try:
            response = self.post('/auth/magic-link/verify', json={'token': token})
            
            if response.status_code == 200:
                data = response.json()
                if 'session_token' in data:
                    self.set_token(data['session_token'])
                return True, data, None
            else:
                error_msg = response.json().get('message', f"Magic link verification failed: {response.status_code}")
                return False, None, error_msg
        except Exception as e:
            return False, None, f"Magic link verification error: {str(e)}"
    
    def validate_license(self) -> Tuple[bool, Optional[Dict[str, Any]], Optional[str]]:
        """Validate license"""
        try:
            response = self.get('/license/validate')
            
            if response.status_code == 200:
                return True, response.json(), None
            else:
                error_msg = response.json().get('message', f"License validation failed: {response.status_code}")
                return False, None, error_msg
        except Exception as e:
            return False, None, f"License validation error: {str(e)}"
    
    def get_license_info(self) -> Tuple[bool, Optional[Dict[str, Any]], Optional[str]]:
        """Get license information"""
        try:
            response = self.get('/license/info')
            
            if response.status_code == 200:
                return True, response.json(), None
            else:
                error_msg = response.json().get('message', f"License info failed: {response.status_code}")
                return False, None, error_msg
        except Exception as e:
            return False, None, f"License info error: {str(e)}"
    
    def register_device(self, device_name: str) -> Tuple[bool, Optional[Dict[str, Any]], Optional[str]]:
        """Register device"""
        try:
            response = self.post('/devices/register', json={'device_name': device_name})
            
            if response.status_code == 200 or response.status_code == 201:
                return True, response.json(), None
            else:
                error_msg = response.json().get('message', f"Device registration failed: {response.status_code}")
                return False, None, error_msg
        except Exception as e:
            return False, None, f"Device registration error: {str(e)}"
    
    def list_devices(self) -> Tuple[bool, Optional[list], Optional[str]]:
        """List devices"""
        try:
            response = self.get('/devices')
            
            if response.status_code == 200:
                return True, response.json(), None
            else:
                error_msg = response.json().get('message', f"Device list failed: {response.status_code}")
                return False, None, error_msg
        except Exception as e:
            return False, None, f"Device list error: {str(e)}"
    
    def delete_device(self, device_id: int) -> Tuple[bool, Optional[str]]:
        """Delete device"""
        try:
            response = self.delete(f'/devices/{device_id}')
            
            if response.status_code == 200 or response.status_code == 204:
                return True, None
            else:
                error_msg = response.json().get('message', f"Device deletion failed: {response.status_code}")
                return False, error_msg
        except Exception as e:
            return False, f"Device deletion error: {str(e)}"
